numba/cuda/cudadecl.py

```diff
@@ -361,6 +361,39 @@ class Cuda_fp16_binary(ConcreteTemplate):
 Cuda_hneg = _genfp16_unary(cuda.fp16.hneg)
 Cuda_habs = _genfp16_unary(cuda.fp16.habs)
 
+hsin_device = declare_device_function('hsin_wrapper', types.float16,
```

Comment:
Rather than adding all these at global scope, I think it would save some boilerplate to declare a function in `CudaFp16Template` like:

```python
    def _resolve_wrapped_unary(self, fname):
        decl = declare_device_function(f'{fname}_wrapper',
                                       types.float16,
                                       (types.float16,))
        return types.Function(decl)
```

then use it as shown below for `hsin`.

Reply:
@gmarkall I partially implemented your recommendation: I created two global functions _resolve_wrapped_unary and _resolve_wrapped_binary. I had to implement and use them at global scope because device functions needs to be singletions otherwise you get ambigious errors for multiple singatures.